staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go

```diff
@@ -109,45 +84,7 @@ func (strategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
 		newCRD.Generation = oldCRD.Generation + 1
 	}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceValidation) {
-		newCRD.Spec.Validation = nil
-		oldCRD.Spec.Validation = nil
-		for i := range newCRD.Spec.Versions {
-			newCRD.Spec.Versions[i].Schema = nil
-		}
-		for i := range oldCRD.Spec.Versions {
-			oldCRD.Spec.Versions[i].Schema = nil
-		}
-	}
-	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) {
-		newCRD.Spec.Subresources = nil
-		oldCRD.Spec.Subresources = nil
-		for i := range newCRD.Spec.Versions {
-			newCRD.Spec.Versions[i].Subresources = nil
-		}
-		for i := range oldCRD.Spec.Versions {
-			oldCRD.Spec.Versions[i].Subresources = nil
-		}
-	}
-
-	// On UPDATE, if the CustomResourceWebhookConversion feature gate is off, we auto-clear
-	// the per-version fields if the old CRD doesn't use per-version fields already.
-	// This is to be consistent with the other built-in types, as the apiserver drops unknown
-	// fields. If the old CRD already uses per-version fields, the CRD is allowed to continue
-	// use per-version fields.
-	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceWebhookConversion) &&
-		!hasPerVersionField(oldCRD.Spec.Versions) {
```

Comment:
this might be problematic uncovered by testing. if we enable `validation`/`subresources` feature, adding `schema`/`subresources` to an existing CRD's versions will be auto-cleared since that `webhook conversion` is off. or rather, we __have to__ enable `webhook conversion` even when we want to use `validation`/`subresources` only. is it designed?

Reply:
This forbid to set the per-version fields if webhook conversion is disabled. As we didn't have per-version fields in 1.12, this moves this functionality under the webhook-conversion feature gate. I don't see a regression here.

One could argue whether these fields should be independently feature gated, or even considered beta right away. But we didn't do that. So at worst, it's unfortunate that we have to wait for 1.14 to get this feature, but it is not critical.

Reply:
@mbohlool @roycaihw  ^^

Reply:
per-version validation doesn't make sense apart from webhook conversion. per-version validation schemas must not be identical across all versions, and unless you have webhook conversion, that means the different validation could allow in data that would not be round-trippable in other versions.

The fields were documented as gated by the CustomResourceWebhookConversion feature gate, so I don't see an issue.